src/network/message.rs

```diff
@@ -250,28 +258,18 @@ impl<N: Network, E: Environment> Encoder<Message<N, E>> for Message<N, E> {
     type Error = anyhow::Error;
 
     fn encode(&mut self, message: Message<N, E>, dst: &mut BytesMut) -> Result<(), Self::Error> {
-        // Serialize the message into a buffer.
-        let buffer = message.serialize()?;
+        // Prepare the room for the length of the payload.
+        dst.extend_from_slice(&0u32.to_le_bytes());
```

Comment:
Can you help me understand what this line is doing?

Reply:
In order to avoid an intermediate allocation of a vector for the serialized message (whose length could then be known at this point, as it's done right now), we prepare 4B of room in advance, so that we can calculate that length after it has been written to `dst` (after those 4B) and then overwrite those first 4B with the length (line 271).

It preserves the current product of serialization while avoiding allocations.

Reply:
That line is the same as `dst.extend_from_slice(&[0, 0, 0, 0]);`